tools/jest_teamcity_reporter/jest_teamcity_reporter.js

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+'use strict'
+
+const assert = require('assert')
+const path = require('path')
+const util = require('util')
+
+/**
+ * @typedef {import('@jest/types').Config.GlobalConfig} GlobalConfig
+ * @typedef {import('@jest/reporters').Reporter} Reporter
+ * @typedef {import('@jest/reporters').Test} Test
+ * @typedef {import('@jest/reporters').TestResult} TestResult
+ */
+
+/** @implements {Reporter} */
+class TeamcityReporter {
+  /** @type {string} */
+  #suiteName
+
+  /**
+   * @param {GlobalConfig} _globalConfig
+   * @param {Record<string, any>} reporterOptions
+   */
+  constructor(_globalConfig, reporterOptions) {
+    this.#suiteName = reporterOptions.suiteName ?? 'jest'
+  }
+
+  /**
+   * @type {Reporter["onRunStart"]}
```

Comment:
Is this not inferred from the `@implements {Reporter}` on the class? Unlike in Typescript I'd be surprised if this adds anything (there won't be any errors if the implementation doesn't match the typedef), so I'm not sure it's worth adding this.

Same for the other methods - the place these types will be shown/used is when we use an instance of this class somewhere else, and there, these types should be inferred from the `@implements` on the class.

Reply:
This is to support ctrl+click navigation to the exact method in `Reporter` interface. This doesn't add any typecheck capabilities, but makes things clearer.